File: geo3x3/Geo3x3.go
package geo3x3

func Encode(lat, lng float64, level int) string {
	if level < 1 {
		return ""
	}
	res := "E"
	if lng < 0.0 {
		res = "W"
		lng += 180.0
	}
	//lat = 90. - lat // 0:the North Pole,  180:the South Pole
	lat += 90.0 // 0:the South Pole  180:the North Pole
	unit := 180.0
	for i := 1; i < level; i++ {
		unit /= 3.0
		x := int(lng / unit)
		y := int(lat / unit)
		res += string(rune('0' + x + y*3 + 1))
		lng -= float64(x) * unit
		lat -= float64(y) * unit
	}
	return res
}

func Decode(code string) []float64 {
	flg := code[0] == 'W'
	unit := 180.0
	lat := 0.0
	lng := 0.0
	level := 1
	for _, c := range code[1:] {
		if c < '1' || c > '9' {
			break
		}
		n := int(c - '1')
		unit /= 3.0
		lng += float64(n%3) * unit
		lat += float64(n/3) * unit
		level++
	}
	lat += unit / 2.0
	lng += unit / 2.0
	//lat = 90 - lat
	lat -= 90.0
	if flg {
		lng -= 180.0
	}
	return []float64{lat, lng, float64(level), unit}
}

File: geo3x3.ml
open String;;

let encode lat lng level =
    if level < 1 then
        ""
    else
        let (res, lng1) =
            if lng >= 0.0 then
                ("E", lng)
            else
                ("W", lng +. 180.0) in
        let lng2 = lng1 in
        let lat2 = lat +. 90.0 in (* 180:the North Pole,  0:the South Pole *)
        let rec loop i lat lng unit =
            if i >= level then
                ""
            else
                let x = int_of_float (lng /. unit) in
                let y = int_of_float (lat /. unit) in
                let n = x + y * 3 + 1 in
                let lng2 = lng -. (float_of_int x) *. unit in
                let lat2 = lat -. (float_of_int y) *. unit in
                (string_of_int n) ^ (loop (i + 1) lat2 lng2 (unit /. 3.0)) in
        res ^ (loop 1 lat2 lng2 (180.0 /. 3.0));;

let decode code = (* (35., 135., 14, 0.1);; *)
    let clen = String.length code in
    if clen < 1 then
        (0.0, 0.0, 0, 180.0) (* err *)
    else
        let c = code.[0] in
        let (flg, beg) =
            if c = '-' || c = 'W' then
                (true, 1)
            else if c = '+' || c = 'E' then
                (false, 1)
            else
                (false, 0) in
        let rec loop i lat lng level unit =
            if i >= clen then
                (lat, lng, level, unit)
            else
                let unit2 = unit /. 3.0 in
                let n = (int_of_char code.[i]) - 49 in (* ignore err check *)
                let lat2 = lat +. (float_of_int (n / 3)) *. unit2 in
                let lng2 = lng +. (float_of_int (n mod 3)) *. unit2 in
                loop (i + 1) lat2 lng2 (level + 1) unit2 in
        let (lat, lng, level, unit) = loop beg 0.0 0.0 1 180.0 in
        let lat2 = lat +. (unit /. 2.0) in
        let lng2 = lng +. (unit /. 2.0) in
        let lat3 = lat2 -. 90.0 in
        let lng3 = if flg then lng2 -. 180.0 else lng2 in
        (lat3, lng3, level, unit);;
